### PR TITLE
stash-debug: add delete and JSON support

### LINKED ISSUE #20306: Add a `delete` function to the `stash-debug` tool
### Feature request

When migrations fail due to catalog corruption (see #5522), it may be critical to be able to quickly repair the corruption so that the system can be restarted. In this case, use of the `stash-debug` tool would be advisable, avoiding manual updates to the catalog database. However, the `stash-debug` tool presently does not include a `delete` API, which would be necessary to remove objects with corrupted definitions.

This is a request to include `delete` write functionality in `stash-debug` in addition to `edit`. The new function should be given a collection and a key, and remove the item with the given key from the named collection.

### PR DESCRIPTION
**Pull request by mjibson**
Fixes #20306

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

- Best viewed commit-by-commit.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

**Comment by mjibson** [comment]
Example output:

```
    "system_privileges": [
         {
            key: SystemPrivilegesKey {
                grantee: Some(
                    RoleId {
                        value: Some(
                            User(
                                1,
                            ),
                        ),
                    },
                ),
                grantor: Some(
                    RoleId {
                        value: Some(
                            System(
                                1,
                            ),
                        ),
                    },
                ),
            },
            value: SystemPrivilegesValue {
                acl_mode: Some(
                    AclMode {
                        bitflags: 3758096384,
                    },
                ),
            },
            key_json: '{"grantee":{"value":{"User":1}},"grantor":{"value":{"System":1}}}',
            value_json: '{"acl_mode":{"bitflags":3758096384}}',
            timestamp: -9223372036854775805,
            diff: 1,
        },
    ],
```

File: src/stash-debug/src/main.rs [hunk 1]
```diff
@@ -116,12 +116,17 @@ pub struct Args {
 #[derive(Debug, clap::Subcommand)]
 enum Action {
     Dump {
```
Comment (benesch) [inline src/stash-debug/src/main.rs:118]:
```suggestion
    /// Dumps the stash contents to stdout in a human readable format.
    /// Includes JSON for each key and value that can be hand edited and
    /// then passed to the `edit` or `delete` commands.
    Dump {
```

File: src/stash-debug/src/main.rs [hunk 2]
```diff
@@ -116,12 +116,17 @@ pub struct Args {
...
+        /// Write output to specified path. Default stdout.
         target: Option<PathBuf>,
     },
     Edit {
         collection: String,
-        key: Vec<u8>,
-        value: Vec<u8>,
+        key: serde_json::Value,
```
Comment (benesch) [inline src/stash-debug/src/main.rs:124]:
```suggestion
        /// The JSON-encoded key that identifies the item to edit.
        key: serde_json::Value,
```

File: src/stash-debug/src/main.rs [hunk 3]
```diff
@@ -116,12 +116,17 @@ pub struct Args {
...
+        value: serde_json::Value,
```
Comment (benesch) [inline src/stash-debug/src/main.rs:125]:
```suggestion
        /// The new JSON-encoded value for the item.
        value: serde_json::Value,
```

File: src/stash-debug/src/main.rs [hunk 4]
```diff
@@ -116,12 +116,17 @@ pub struct Args {
 #[derive(Debug, clap::Subcommand)]
 enum Action {
     Dump {
+        /// Write output to specified path. Default stdout.
         target: Option<PathBuf>,
     },
     Edit {
         collection: String,
```
Comment (benesch) [inline src/stash-debug/src/main.rs:123]:
```suggestion
        /// The name of the stash collection to edit.
        collection: String,
```

File: src/stash-debug/src/main.rs [hunk 5]
```diff
@@ -116,12 +116,17 @@ pub struct Args {
 #[derive(Debug, clap::Subcommand)]
 enum Action {
     Dump {
+        /// Write output to specified path. Default stdout.
         target: Option<PathBuf>,
     },
     Edit {
```
Comment (benesch) [inline src/stash-debug/src/main.rs:122]:
```suggestion
    /// Edits a single item in a collection in the stash.
    Edit {
```

File: src/stash-debug/src/main.rs [hunk 6]
```diff
@@ -116,12 +116,17 @@ pub struct Args {
...
         collection: String,
-        key: Vec<u8>,
-        value: Vec<u8>,
+        key: serde_json::Value,
+        value: serde_json::Value,
+    },
+    Delete {
+        collection: String,
+        key: serde_json::Value,
```
Comment (benesch) [inline src/stash-debug/src/main.rs:129]:
```suggestion
    /// Deletes a single item in a collection in the stash
    Delete {
        /// The name of the stash collection to edit.
        collection: String,
        /// The JSON-encoded key that identifies the item to delete.
        key: serde_json::Value,
```

**Review by benesch — APPROVED** [review]
This is perfect, Matt. Thank you so much.

I wrote up a playbook for using this tool here: https://www.notion.so/materialize/Debugging-Materialize-Cloud-0788c1fe692b4acf9d247345e7b5346e?pvs=4#f2f12d736eae427c8892b9e561bb0be9

LGTM. Just left you some suggestions with more useful help output.

**Pull request merged**